chain/network/src/ibf.rs

```diff
@@ -0,0 +1,255 @@
+    }
+
+    /// Calculate minimum parameter 'k', such that given IBF has at least 'capacity` elements.
+    fn calculate_k(capacity: usize) -> i32 {
+        let mut k = 0;
+        // In order make computation of indexes not require doing division, we allocate
+        // 2^k + NUM_HASHES - 1 hashes. This allows us to use compute indexes using bits.
+        while (1 << k) + NUM_HASHES - 1 < capacity {
+            k += 1;
+        }
+        k
+    }
+
+    /// Add element to the set. It's up to the caller to make sure this method is called only if
+    /// the item is not in the set.
+    pub fn add(&mut self, elem: u64) {
+        self.insert(elem)
+    }
+
+    /// Remove element from the set. It's up to the caller to make sure this method is called only
+    /// if the item is in the set.
+    /// It's also worth noting that add/remove functions are identical.
+    pub fn remove(&mut self, elem: u64) {
+        self.insert(elem)
+    }
+
+    /// Compute hash of element
+    fn compute_hash(&self, elem: u64) -> u64 {
+        let mut h = self.hasher.clone();
+        h.write_u64(elem);
+        h.finish()
+    }
+
+    /// Add element to IBF, which involves updating values of 3 boxes.
+    fn insert(&mut self, elem: u64) {
+        self.insert_value(elem);
+    }
+
+    /// Merge two sets together, the result will be IbfSet representing symmetric difference
+    /// between those sets.
+    pub fn merge(&mut self, rhs_data: &[IbfBox], rhs_seed: u64) -> bool {
+        if self.data.len() != rhs_data.len() || self.seed != rhs_seed {
+            error!(
+                "failed to merge len: {} {} seed: {} {}",
+                self.data.len(),
+                rhs_data.len(),
+                self.seed,
+                rhs_seed
+            );
+            return false;
+        }
+        for (lhs, rhs)  in self.data.iter_mut().zip(rhs_data) {
+            lhs.merge(rhs)
+        }
+        true
+    }
+
+    #[cfg(test)]
+    fn recover(&mut self) -> Result<Vec<u64>, &'static str> {
+        let (result, difference) = self.try_recover();
+
+        if difference != 0 {
+            for i in 0..self.data.len() {
+                if self.data[i].xor_elem != 0 {
+                    println!(
+                        "{} {:?} {}",
+                        i,
+                        self.data[i],
+                        self.compute_hash(self.data[i].xor_elem)
+                    );
+                }
+            }
+            return Err("unable to recover result");
+        }
+        return Ok(result);
+    }
+
+    /// Try to recover elements inserted into IBF.
+    ///
+    /// Returns list of recovered elements and number of boxes that were not recoverable.
+    pub fn try_recover(&mut self) -> (Vec<u64>, u64) {
+        let mut result = Vec::with_capacity(self.data.len());
+        let mut to_check = Vec::new();
+        for i in 0..self.data.len() {
+            to_check.push(i);
```

Comment:
Why don't we allocate `to_check` in advance similar to how `result` is initialized?